packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,553 @@
+      selected: new _TransformedRect(selected),
+      tooltip: selected.toString(),
+      candidates: candidates,
+    );
+
+    if (state != _lastState) {
+      _lastState = state;
+      _picture = _buildPicture(state);
+    }
+    builder.addPicture(layerOffset, _picture);
+  }
+
+  ui.Picture _buildPicture(_InspectorOverlayRenderState state) {
+    final ui.PictureRecorder recorder = new ui.PictureRecorder();
+    final Canvas canvas = new Canvas(recorder, state.overlayRect);
+    final Size size = state.overlayRect.size;
+
+    final Paint fillPaint = new Paint()
+      ..style = PaintingStyle.fill
+      ..color = _kHighlightedRenderObjectFillColor;
+
+    final Paint borderPaint = new Paint()
+      ..style = PaintingStyle.stroke
+      ..strokeWidth = 1.0
+      ..color = _kHighlightedRenderObjectBorderColor;
+
+    // Highlight the selected renderObject.
+    final Rect selectedPaintRect = state.selected.rect.deflate(0.5);
+    canvas
+      ..save()
+      ..transform(state.selected.transform.storage)
+      ..drawRect(selectedPaintRect, fillPaint)
+      ..drawRect(selectedPaintRect, borderPaint)
+      ..restore();
+
+    // Show all other candidate possibly selected elements. This helps selecting
+    // render objects by selecting the edge of the bounding box shows all
+    // elements the user could toggle the selection between.
+    for (_TransformedRect transformedRect in state.candidates) {
+      canvas
+        ..save()
+        ..transform(transformedRect.transform.storage)
+        ..drawRect(transformedRect.rect.deflate(0.5), borderPaint)
+        ..restore();
+    }
+
+    final Rect targetRect = MatrixUtils.transformRect(
+        state.selected.transform, state.selected.rect);
+    final Offset target = new Offset(targetRect.left, targetRect.center.dy);
+    final double offsetFromWidget = 9.0;
+    final double verticalOffset = (targetRect.height) / 2 + offsetFromWidget;
+
+    _paintDescription(canvas, state.tooltip, target, verticalOffset, size, targetRect);
+
+    // TODO(jacobr): provide an option to perform a debug paint of just the
+    // selected widget.
+    return recorder.endRecording();
+  }
+
+  void _paintDescription(Canvas canvas, String message, Offset target,
+      double verticalOffset, Size size, Rect targetRect) {
+    canvas.save();
+    final double maxWidth = size.width - 2 * (_kScreenEdgeMargin + _kTooltipPadding);
+    if (_textPainter == null || _textPainter.text.text != message || _textPainterMaxWidth != maxWidth) {
+      _textPainterMaxWidth = maxWidth;
+      _textPainter = new TextPainter()
+        ..maxLines = _kMaxTooltipLines
+        ..ellipsis = '...'
+        ..text = new TextSpan(style: _messageStyle, text: message)
+        ..layout(maxWidth: maxWidth);
+    }
+
+    final Size tooltipSize = _textPainter.size + const Offset(_kTooltipPadding * 2, _kTooltipPadding * 2);
+    final TooltipPositionDelegate tooltipPositionDelegate = new TooltipPositionDelegate(
```

Comment:
Creating a custom layout delegate when you really just want to call a function doesn't really make sense. Maybe we should instead leave the custom layout delegate where it was, and factor out the math into a reusable function in the painting layer?

I hadn't realised you weren't actually using a custom layout widget here.

Adam's suggestion of actually using a real pipeline seems really solid. It would let you remove all these hacks and just implement a real widget UI with all of the layout primitives and so on.